File: 백준/Silver/9012. 괄호/괄호.py
import sys
from collections import deque
input = sys.stdin.readline

n = int(input())
bracket_list = [list(input().strip()) for _ in range(n)]

def check(string_list):
    queue = deque(string_list)
    left = 0
    right = 0
    # print(queue)

    while queue:
        item = queue.popleft()
        if item == '(':
            left += 1
        else:
            right += 1
        if right > left:
            return 'NO'
    return 'YES' if left == right else 'NO'

for arr in bracket_list:
    print(check(arr))

File: 백준/Silver/9012. 괄호/괄호.js
let input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const T = input[0];
const inputList = input.slice(1);

const check = (parenthesesList) => {
    let result = 0;
    const list = parenthesesList.split('');
    for (let item of list) {
        if (item === '(') result += 1;
        else if (item === ')') result -= 1;
        if (result < 0) return 'NO';
    }
    return result === 0 ? 'YES' : 'NO';
};

inputList.forEach((list) => {
    console.log(check(list));
});
